tools/please_pex/pex/pex.go

```diff
@@ -0,0 +1,184 @@
+// we were ultimately not using pex for much at construction time and
+// we already have most of what we need in Go via jarcat.
+package pex
+
+import (
+	"bytes"
+	"embed"
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+const testRunnersDir = "test_runners"
+const debuggersDir = "debuggers"
+
+type Debugger string
+
+const (
+	Pdb     Debugger = "pdb"
+	Debugpy Debugger = "debugpy"
+)
+
+//go:embed *.py
+//go:embed test_runners/*.py
+//go:embed debuggers/*.py
+var files embed.FS
+
+// A Writer implements writing a .pex file in various steps.
+type Writer struct {
+	zipSafe          bool
+	noSite           bool
+	shebang          string
+	realEntryPoint   string
+	pexStamp         string
+	testSrcs         []string
+	includeLibs      []string
+	testRunner       string
+	customTestRunner string
+	debugger         string
+}
+
+// NewWriter constructs a new Writer.
+func NewWriter(entryPoint, interpreter, options, stamp string, zipSafe, noSite bool) *Writer {
+	pw := &Writer{
+		zipSafe:        zipSafe,
+		noSite:         noSite,
+		realEntryPoint: toPythonPath(entryPoint),
+		pexStamp:       stamp,
+	}
+	pw.SetShebang(interpreter, options)
+	return pw
+}
+
+// SetShebang sets the leading shebang that will be written to the file.
+func (pw *Writer) SetShebang(shebang string, options string) {
+	shebang = strings.TrimSpace(fmt.Sprintf("%s %s", shebang, options))
+	if !path.IsAbs(shebang) {
+		shebang = "/usr/bin/env " + shebang
+	}
+	if pw.noSite {
+		// In many environments shebangs cannot have more than one argument; we can work around
+		// that by treating it as a shell script.
+		if strings.Contains(shebang, " ") {
+			shebang = "#!/bin/sh\nexec " + shebang + ` -S $0 "$@"`
+		} else {
+			shebang += " -S"
+		}
+	}
+	if !strings.HasPrefix(shebang, "#") {
+		shebang = "#!" + shebang
+	}
+	pw.shebang = shebang + "\n"
+}
+
+// SetTest sets this Writer to write tests using the given sources.
+// This overrides the entry point given earlier.
+func (pw *Writer) SetTest(srcs []string, testRunner string) {
+	pw.realEntryPoint = "pex_test_main"
+	pw.testSrcs = srcs
+
+	switch testRunner {
+	case "pytest":
+		pw.testRunner = filepath.Join(testRunnersDir, "pytest.py")
+	case "behave":
+		pw.testRunner = filepath.Join(testRunnersDir, "behave.py")
+	case "unittest":
+		pw.testRunner = filepath.Join(testRunnersDir, "unittest.py")
+	default:
+		if !strings.ContainsRune(testRunner, '.') {
+			panic("Custom test runner '" + testRunner + "' is invalid; must contain at least one dot")
+		}
+		pw.testRunner = filepath.Join(testRunnersDir, "custom.py")
+		pw.customTestRunner = testRunner
+	}
+}
+
+func (pw *Writer) SetDebugger(debugger Debugger) {
+	pw.pexStamp = "debug"
+
+	switch debugger {
+	case "pdb":
+		pw.debugger = filepath.Join(debuggersDir, "pdb.py")
+	case "debugpy":
+		pw.debugger = filepath.Join(debuggersDir, "debugpy.py")
+		pw.includeLibs = append(pw.includeLibs, ".bootstrap/debugpy")
```

Comment:
I think this isn't needed anymore, if debugpy now lives in third_party?

Reply:
Actually `includeLibs` doesn't really do anything anymore... I'm wondering if debugpy is actually being pulled in anywhere now

Reply:
Yes, it should be included now as a dep. Check my latest changes on https://github.com/thought-machine/please/pull/2213 which might point you in the right direction